coverage.py

```diff
@@ -174,7 +189,12 @@ def operation(self):
         return self.OPERATION
 
     def run(self):
-        self.run_once_and_update_timestamp()
+        start = datetime.datetime.utcnow()
+        result = self.run_once_and_update_timestamp()
+
+        result = result or CoverageProviderProgress()
+        self.finalize_timestampdata(result, start=start)
+        return result
```

Comment:
This pattern will also show up in Monitor and Script. The generic method calls the "do something" method, expects a TimestampData back, and writes the TimestampData to the database.